### PR TITLE
AirfRANS: Pressure-weighted loss (20x) at 4L/256d + lr=3e-4

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The pressure-weight breakthrough (0.00435) used lr=3e-4 + T_max=10 at 3L/192d. We test the SAME hyperparameters at 4L/256d to directly measure the architecture scaling effect on pressure-weighted training. This isolates whether 4L/256d provides additional capacity benefits when gradient allocation is already optimized by pressure weighting.

## Code Change Required
Same as edward's experiment — add `--pressure-loss-weight` support. Reference nami's implementation on branch `nami/airfrans-pressure-weighted-loss`.

1. Add argument: `parser.add_argument('--pressure-loss-weight', type=float, default=1.0, help='Weight multiplier for pressure channel in loss')`
2. Weight the pressure channel MSE by this factor in the loss computation. Pressure is typically channel index 2 (Ux=0, Uy=1, p=2, nut=3).

## Training Command
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 3e-4 \
  --cosine-t-max 10 \
  --pressure-loss-weight 20.0 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --seed 789 \
  --agent haku \
  --wandb-name "haku/airfrans-4L256d-pressure20x-lr3e4" \
  --wandb-group "airfrans-pressure-weight"
```

Note: Using seed=789 (best seed from multi-seed characterization) and nami's exact lr=3e-4 + T_max=10.

Report val_primary/surface_mse, best epoch, per-channel breakdown, and total epochs completed.

## Baseline
Current best: **0.007264** (4L/256d golden config, NO pressure weight)
Pressure-weight reference: **0.00435** (3L/192d, lr=3e-4, T_max=10, 20x) — W&B toeli7xw
External target: **0.0043**

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT haku:

## Results

### Key Metrics (Best Epoch 58 of 80)

| Metric | Baseline (4L/256d) | This Run (4L/256d + 20x pressure) | Pressure Ref (3L/192d + 20x) |
|--------|--------------------|------------------------------------|------------------------------|
| **surface_mse** | **0.007264** | 0.02341 | **0.00435** |
| surface_mse_Ux | — | 2.80e-05 | — |
| surface_mse_Uy | — | 5.15e-06 | — |
| surface_mse_p | — | 0.09359 | — |
| surface_mse_nut | — | 2.32e-05 | — |
| volume_mse | — | 0.1712 | — |

### Summary

**The experiment failed.** 20x pressure-weighted loss at 4L/256d achieved best surface_mse = 0.02341 at epoch 58, which is **3.2x worse than the unweighted baseline (0.007264)** and **5.4x worse than the 3L/192d pressure-weight reference (0.00435)**.

### Training Dynamics

The 20x pressure weight caused severe training instability characterized by:

1. **Violent oscillations**: Each cosine annealing LR restart caused massive surface_mse spikes (e.g., epoch 63: 0.467, epoch 80: 0.327), far worse than typical restart perturbation with unweighted loss
2. **Catastrophic divergence after epoch 63**: The model never recovered from the cycle 7 LR restart. Surface_mse oscillated between 0.09-0.33 for 17 epochs, never approaching its best of 0.023
3. **Extremely high grad norms**: Mean grad norms were 300-500+ throughout training (vs typical ~10-50 for unweighted runs), indicating the 20x pressure weight created a very steep loss landscape
4. **Slow early convergence**: Took 40+ epochs to reach surface_mse=0.058, whereas the baseline reached 0.007 by epoch 40

### Cycle-over-cycle progression (best surface_mse per T_max=10 cycle)

| Cycle | Epochs | Best surface_mse | Improvement |
|-------|--------|------------------|-------------|
| 1 | 1-10 | 0.244 | — |
| 2 | 11-20 | 0.210 | 14% |
| 3 | 21-30 | 0.103 | 51% |
| 4 | 31-40 | 0.058 | 44% |
| 5 | 41-50 | 0.029 | 50% |
| 6 | 51-60 | 0.023 | 21% |
| 7 | 61-70 | 0.093 | **-304% (diverged)** |
| 8 | 71-80 | 0.109 | **no recovery** |

Training was terminated at epoch 80 after the model catastrophically diverged and showed no signs of recovery.

### Why It Failed

The 20x pressure weight at 4L/256d creates a fundamentally different optimization landscape than at 3L/192d:

1. **Capacity mismatch**: The larger 4L/256d model has more parameters to destabilize. The 20x pressure weight amplifies gradients through more layers, causing larger parameter updates that are destructive at high LR phases
2. **Gradient norm explosion**: Mean grad norms of 300-500 suggest the effective learning rate was far too high for this loss weighting. The standard grad_clip=1.0 was insufficient to tame the pressure-amplified gradients
3. **Cosine annealing incompatibility**: The T_max=10 cycling was too aggressive for 20x-weighted loss at this model size — each LR peak catastrophically disrupted learned representations

### Experiment Details

- **Command**: `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python -u train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 3e-4 --cosine-t-max 10 --pressure-loss-weight 20.0 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --seed 789 --agent haku --wandb-name "haku/airfrans-4L256d-pressure20x-lr3e4" --wandb-group "airfrans-pressure-weight"`
- **W&B run**: `zb638gs3`
- **Peak memory**: ~8.4 GB VRAM
- **Epochs completed**: 80 (terminated early due to catastrophic divergence)

### Suggested Follow-ups

1. **Lower pressure weight at 4L/256d**: Try 5x or 10x instead of 20x — the larger model may need a gentler pressure signal
2. **Lower LR with 20x weight**: Try lr=1e-4 or 5e-5 to compensate for the amplified gradients
3. **Stronger grad clipping**: Try grad_clip=0.5 or 0.1 to tame the 300-500 grad norms
4. **Longer T_max**: T_max=20 or 30 would reduce the severity of LR restarts and may prevent the catastrophic oscillations

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2802 because ADVISOR: Closing — 20x pressure-weighted loss at 4L/256d is catastrophically harmful (0.02341 vs baseline 0.00277, 8.4x regression). Root cause: 4L/256d amplifies the pressure gradient signal through 4 layers, making the effective weight ~80x when accounting for depth. Mean grad norms 300-500 confirm the loss landscape is fundamentally distorted. The 20x weight works at 3L/192d because the shallower gradient path is more stable. At 4L/256d, pressure weight must be much lower (try 5x or 10x). Assigning a new experiment next..
